lib/internal/streams/operators.js

```diff
@@ -218,11 +219,57 @@ async function * filter(fn, options) {
   yield* this.map(filterFn, options);
 }
 
+// Specific to provide better error to reduce since the argument is only
+// missing if the stream has no items in it - but the code is still appropriate
+class ReduceAwareErrMissingArgs extends ERR_MISSING_ARGS {
+  constructor() {
+    super('reduce');
+    this.message = 'Reduce of an empty stream requires an initial value';
+  }
+}
+
+async function reduce(reducer, initialValue, options) {
+  if (typeof reducer !== 'function') {
+    throw new ERR_INVALID_ARG_TYPE(
+      'reducer', ['Function', 'AsyncFunction'], reducer);
+  }
+  let hasInitialValue = arguments.length > 1;
+  if (options?.signal?.aborted) {
+    const err = new AbortError(undefined, { cause: options.signal.reason });
+    this.once('error', () => {}); // The error is already propagated
+    this.destroy(err);
```

Comment:
This should ideally wait for `destroy()` to complete, i.e. the `close` event. If the stream is not fully closed before we return from `reduce()`, it can lead to (userland) issues like lock contention. For reference, `pipeline()` and `finished()` had the same problem: https://github.com/nodejs/node/pull/32158

Reply:
Good point, is this a guarantee we make? 

APIs I've worked on in the past only make the "eventually close" guarantee (like `addAbortSignal` or passing the signal to the stream) and if users want to check if the stream has ended it's up to them to check the stream itself

Reply:
sure

Reply:
finished is currently not internal which means I (probably) can't import it here - should I move it around (add the code to the eos code) or use eos directly?

Reply:
I'll just move it to eos and re-export it from promises.js